analyzer/src/main/kotlin/managers/Npm.kt

```diff
@@ -389,67 +371,87 @@ open class Npm(
+        buildTree(moduleDir, scopes)!!.dependencies
+
     private fun buildTree(
-        rootModulesDir: File, startModulesDir: File, moduleName: String, packages: Map<String, Package>,
-        dependencyBranch: List<String> = emptyList()
+        moduleDir: File,
+        scopes: Set<String>,
+        ancestorModuleDirs: List<File> = emptyList(),
+        ancestorModuleIds: Set<Identifier> = emptySet(),
+        packageType: String = managerName
     ): PackageReference? {
-        log.debug { "Building dependency tree for '$moduleName' from directory '$startModulesDir'." }
-
-        val packageFile = startModulesDir.resolve(moduleName).resolve("package.json")
+        val moduleInfo = getModuleInfo(moduleDir, scopes)
+        val dependencies = sortedSetOf<PackageReference>()
+        val moduleId = splitNamespaceAndName(moduleInfo.name).let { (namespace, name) ->
+            Identifier(packageType, namespace, name, moduleInfo.version)
+        }
 
-        if (packageFile.isFile) {
-            log.debug { "Found package file for module '$moduleName' at '$packageFile'." }
+        if (ancestorModuleIds.contains(moduleId)) {
+            val cycle = ancestorModuleIds.toList().let {
+                val cycleStartIndex = it.indexOf(moduleId)
+                it.subList(cycleStartIndex, it.size) + moduleId
+            }.joinToString(" -> ")
 
-            val packageJson = jsonMapper.readTree(packageFile)
-            val rawName = packageJson["name"].textValue()
-            val version = packageJson["version"].textValue()
-            val identifier = "$rawName@$version"
+            log.debug { "Not adding dependency '$moduleId' to avoid cycle: $cycle." }
+            return null
+        }
 
-            if (identifier in dependencyBranch) {
-                log.debug {
-                    "Not adding circular dependency '$identifier' to the tree, it is already on this branch of the " +
-                            "dependency tree: ${dependencyBranch.joinToString(" -> ")}."
-                }
-                return null
-            }
+        log.debug { "Building dependency tree for '${moduleInfo.name}' from directory '$moduleDir'." }
 
-            val dependencies = sortedSetOf<PackageReference>()
+        val pathToRoot = listOf(moduleDir) + ancestorModuleDirs
+        moduleInfo.dependencyNames.forEach { dependencyName ->
+            val dependencyModuleDirPath = findDependencyModuleDir(dependencyName, pathToRoot)
 
-            val dependencyNames = mutableSetOf<String>()
-            packageJson["dependencies"]?.fieldNames()?.forEach { dependencyNames += it }
-            packageJson["optionalDependencies"]?.fieldNames()?.forEach { dependencyNames += it }
+            if (dependencyModuleDirPath.isNotEmpty()) {
+                val dependencyModuleDir = dependencyModuleDirPath.first()
+                log.debug { "Found module dir for '$dependencyName' at '$dependencyModuleDir'." }
```

Comment:
Does this print the paths relative to the current module or relative to the root package.json? I would prefer the latter for easier debugging.

Reply:
This prints the absolute path to the module directory, where the dependency is installed.